VMEncryption/volume-notif-svc/UdevVolNotif.c

```diff
@@ -0,0 +1,387 @@
+void invoke_ade(const char* path){
+    ade_status = AdeStatus::ADE_RUNNING;
+    pid_t pAde = fork();
+    if(pAde==0){
+            // in case of child process.
+            custom_log("child process %d created, running ADE", getpid());
+            custom_log("changing directory to %s", path);
+            chdir(path);
+            execl(          "extension_shim.sh",
+                            "extension_shim.sh",
+                            "-c",
+                            "main/handle.py --enable --vnscall",
+                            NULL);
+            int err = errno;
+            custom_log("Failed to run process! error: %d",err);
+            exit(1);
+    }
+    if(pAde>0){
+       pid_t childPid = wait(NULL);
+       custom_log("child process %d is completed",childPid);
+    }
+    //checking if daemon is still running
+    pid_t pDaemon = fork();
+    if(pDaemon==0){
+            custom_log("child process %d created, checking daemon", getpid());
+            custom_log("changing directory to %s", path);
+            chdir(path);
+            execl(          "ade_daemon_delay.sh",
+                            "ade_daemon_delay.sh",
+                            "extension_shim.sh",
+                            NULL);
+            int err = errno;
+            custom_log("Failed to run process! error: %d",err);
+            exit(1);
+
+    }
+    if(pDaemon>0){
+       pid_t childPid = wait(NULL);
+       custom_log("child process %d is completed",childPid);
+    }
+    ade_status = AdeStatus::ADE_FINISHED;
+    ade_finished = time(NULL);
+}
+
+int is_device_crypted_from_syspath(const char* syspath){
+        struct udev *udev;
+        int ret = 0;
+        udev = udev_new();
+        struct udev_device* dev = udev_device_new_from_syspath(udev,syspath);
+        const char* fs_usage = udev_device_get_property_value(dev,"ID_FS_USAGE");
+        custom_log("get_dev_fsUsage_from_syspath: syspath %s usage status is %s",syspath, fs_usage);
+        if(fs_usage!=NULL && strcmp(fs_usage,"crypto")==0) ret = 1;
+        udev_device_unref(dev);
+        udev_unref(udev);
+        return ret;
+}
+
+bool is_devnode_added(std::deque<device>&dq, const char* syspath, const char* action){
+        for(auto it:dq)
+                if(             strcmp(it.syspath.c_str(),syspath)==0 &&
+                                strcmp(it.action.c_str(),action)==0) return true;
+        return false;
+}
+
+void cleanCryptedDevFromList(std::deque<device>&dq){
+        using itdeque=std::deque<device>::iterator;
+        std::vector<itdeque> cryptedDevices;
+        //removing crypted changed devices,
+        for(itdeque it=dq.begin(); it!=dq.end(); it++){
+                if(             it->action =="change" &&
+                                is_device_crypted_from_syspath(it->syspath.c_str())){
+                        cryptedDevices.push_back(it);
+                }
+        }
+        //remove previously add crypted devices, which are mounted now.
```

Comment:
errata: "which are unlocked now", 
assuming status change from add->change for crypted device is inferring ADE unlock to those devices.